rfcs/taskforce/0002-network-utility.md

```diff
@@ -0,0 +1,203 @@
+- The internal data structures for storing these keypairs, genesis configuration and distribution are not well defined and could be more robust
+- Tools for deployment and monitoring are seperate and these could be packaged in one simple workflow that uses the same underlying functionality
+
+## Proposal
+
+All of these features should be included in a simple command line tool that also exposes a TypeScript API. This service will be written in Reason with TypeScript bindings.
+
+Since there may be some initial configuration required for the operation of a network, the following command will take care of setting up all the required services (google-cloud-sdk, aws, etc.)
+
+```
+coda-network init
+```
+
+### Keypair Management
+
+Keypairs have a private key and a corresponding public key. For new keypair generation, we can use the new client-sdk which is a much lighter dependency and has a clean JavaScript interface.
+
+However there are also existing keys which we'll need to be able to reference by public key alone since we don't control the private keys. An example of this would be community owned wallets for inclusion in the genesis ledger. Since this distinction can be made at a higher level, we'll keep the definition of a keypair to mean a public key for which we also have th private key.
+
+```
+coda-network keypair create
+```
+
+This command will generate a new keypair and save it locally.
+
+```
+coda-network keypair upload <PUBLIC_KEY>
+```
+
+This command will upload a locally stored keypair to an S3 bucket for distribution.
+
+```
+coda-network keypair download <PUBLIC_KEY>
+```
+
+This will attempt to download a given keypair from S3 if the keypair exists.
+
+### Keyset Management
+
+Keysets are simply a collection of keypairs. We use these for categorizing different groups of keys that will be used in the lifecycle of a network. For each keypair in a keyset, we MAY or MAY NOT have the corresponding private keys. For each keyset, we may want to allocate a different amount of funds to their corresponding accounts in a genesis ledger.
+
+The datastructure for a given keyset is much simpler since it's essentially a list public keys. As such we can simply maintain a simple CSV style document that contains all the public keys included in a given keyset using the name of the keyset for the filename. 
+
+```
+coda-network keyset create <NAME>
+```
+
+This simply creates a new keyset that keys can be included in.
+
+```
+coda-network keyset add <NAME> <PUBLIC_KEY> ...
+```
+
+This command would add a public key to a keyset.
+
+```
+coda-network keyset remove <NAME> <PUBLIC_KEY> ...
+```
+
+This command would remove a public key to a keyset.
+
+```
+coda-network keyset [ls|list]
+```
+
+This command will show all keysets that are stored locally on the machine.
+
+```
+coda-network keyset show <NAME>
+```
+
+This command displays all public keys included in a given keyset.
+
+```
+coda-network keyset import <KEYSET_NAME> <CSV_FILENAME>
```

Comment:
Can you specify the schema for the CSV that we'll be expecting? Is it just one column with public keys? I think we should also (optionally?) include "nicknames" for the public keys. This could then be used to consume the discord-handle <-> public-key mapping.

Reply:
Moreover, we can also name our public keys something that has more meaning: "faucet" or "large-o1-block-producer1" for example

Reply:
Great suggestion.